src/autopas/utils/SimilarityFunctions.h

```diff
@@ -8,106 +8,111 @@
-      index[i] = particleLocation[i] / cellLength;
-    }
-    const size_t cellIndex = autopas::utils::ThreeDimensionalMapping::threeToOneD(index, cellsPerDimension);
-    particlesPerCell[cellIndex] += 1;
-    // calculate the size of the current cell
-    allVolumes[cellIndex] = 1;
-    for (size_t i = 0; i < cellsPerDimension.size(); ++i) {
-      // the last cell layer has a special size
-      if (index[i] == cellsPerDimension[i] - 1) {
-        allVolumes[cellIndex] *= outerCellSizePerDimension[i];
-      } else {
-        allVolumes[cellIndex] *= cellLength;
-      }
-    }
+
+    const auto binIndex3d =
+        autopas::utils::ArrayMath::floorToInt((particleLocation - container.getBoxMin()) / binDimensions);
+    const auto binIndex1d = autopas::utils::ThreeDimensionalMapping::threeToOneD(
+        binIndex3d, {numberOfBinsPerDim, numberOfBinsPerDim, numberOfBinsPerDim});
+
+    particlesPerBin[binIndex1d] += 1;
   }
 
-  // calculate density for each cell
+  // calculate density for each bin and track max density
   double maxDensity{0.};
-  std::vector<double> densityPerCell(numberOfCells, 0.0);
-  for (size_t i = 0; i < particlesPerCell.size(); i++) {
-    densityPerCell[i] =
-        (allVolumes[i] == 0) ? 0 : (particlesPerCell[i] / allVolumes[i]);  // make sure there is no division of zero
-    if (densityPerCell[i] > maxDensity) {
-      maxDensity = densityPerCell[i];
-    }
+  std::vector<double> densityPerBin{};
+  densityPerBin.reserve(numberOfBins);
+  for (size_t i = 0; i < numberOfBins; i++) {
+    densityPerBin[i] = static_cast<double>(particlesPerBin[i]) / binVolume;
+    maxDensity = std::max(maxDensity, densityPerBin[i]);
   }
 
-  if (maxDensity < 0.0)
-    throw std::runtime_error("maxDensity can never be smaller than 0.0, but is:" + std::to_string(maxDensity));
-
+  if (maxDensity < 0.0) {
+    utils::ExceptionHandler::exception(
+        "calculateHomogeneityAndMaxDensity(): maxDensity can never be smaller than 0.0, but is: {}", maxDensity);
+  }
   // get mean and reserve variable for densityVariance
-  const double densityMean = numberOfParticles / volume;
-  double densityVariance = 0.0;
+  const double densityMean = numberOfParticles / domainVolume;
 
-  // calculate densityVariance
-  for (size_t r = 0; r < densityPerCell.size(); ++r) {
-    double distance = densityPerCell[r] - densityMean;
-    densityVariance += (distance * distance / densityPerCell.size());
+  double densityVariance{0.};
+  for (size_t i = 0; i < numberOfBins; ++i) {
+    const double densityDifference = densityPerBin[i] - densityMean;
+    densityVariance += (densityDifference * densityDifference);
   }
 
+  /*
+    double densityVariance = std::transform_reduce(
+          densityPerBin.begin(), densityPerBin.end(), 0.0,
+          std::plus<>(),
+          [densityMean](double density) {
+              double densityDifference = density - densityMean;
+              return densityDifference * densityDifference;
+          }
+      );
```

Comment:
somehow `transform_reduce` is just returning 0., and I was unable to figure out why. So, just for now, I commented it out and went back to for-loop. But this needs to be fixed.